crypto/src/swap.rs

```diff
@@ -26,10 +56,129 @@ pub struct SwapPlaintext {
-        // TODO: implement
-        todo!()
+    // Create a new hash based on the ephemeral public key and shared secret suitable for use as a key for symmetric encryption.
+    //
+    // Implementing this way allows recovery of all swap plaintexts via the seed phrase.
+    //
+    // Theoretically, if a paranoid user did want to achieve forward secrecy, they could choose to encrypt
+    // nonsense bytes as the swap plaintext as the swap ciphertext does not need to be valid for the
+    // swap to succeed, however this is unsupported by the official client.
+    fn derive_symmetric_key(
+        shared_secret: &ka::SharedSecret,
+        epk: &ka::Public,
+    ) -> blake2b_simd::Hash {
+        let mut kdf_params = blake2b_simd::Params::new();
+        kdf_params.hash_length(32);
+        let mut kdf = kdf_params.to_state();
+        kdf.update(&shared_secret.0);
+        kdf.update(&epk.0);
+
+        kdf.finalize()
+    }
+
+    pub fn diversified_generator(&self) -> decaf377::Element {
+        self.b_d
+    }
+
+    pub fn transmission_key(&self) -> ka::Public {
+        self.pk_d
+    }
+
+    /// Use Blake2b-256 to derive an encryption key `ock` from the OVK and public fields.
+    pub fn derive_ock(ovk: &OutgoingViewingKey, epk: &ka::Public) -> blake2b_simd::Hash {
+        // let cv_bytes: [u8; 32] = cv.into();
+        // let cm_bytes: [u8; 32] = cm.into();
+
+        let mut kdf_params = blake2b_simd::Params::new();
+        kdf_params.hash_length(32);
+        let mut kdf = kdf_params.to_state();
+        kdf.update(&ovk.0);
+        // TODO: should we be using the public fields e.g. t1, t2, trading_pair here?
+        // Note implementation uses value commitments...
+        // kdf.update(&cv_bytes);
+        // kdf.update(&cm_bytes);
+        kdf.update(&epk.0);
+
+        kdf.finalize()
+    }
+
+    /// Generate encrypted outgoing cipher key for use with this swap.
+    pub fn encrypt_key(
+        &self,
+        esk: &ka::Secret,
+        ovk: &OutgoingViewingKey,
+    ) -> [u8; OVK_WRAPPED_LEN_BYTES] {
+        let epk = esk.diversified_public(&self.diversified_generator());
+        let kdf_output = SwapPlaintext::derive_ock(ovk, &epk);
+
+        let ock = Key::from_slice(kdf_output.as_bytes());
+
+        let mut op = Vec::new();
+        op.extend_from_slice(&self.transmission_key().0);
+        op.extend_from_slice(&esk.to_bytes());
+
+        let cipher = ChaCha20Poly1305::new(ock);
+
+        // Note: Here we use the same nonce as swap encryption, however the keys are different.
+        // For swap encryption we derive a symmetric key from the shared secret and epk.
+        // However, for encrypting the outgoing cipher key, we derive a symmetric key from the
+        // sender's OVK, and the epk. Since the keys are
+        // different, it is safe to use the same nonce.
+        //
+        // References:
+        // * Section 5.4.3 of the ZCash protocol spec
+        // * Section 2.3 RFC 7539
```

Comment:
I think we want to use the same empheral symmetric key for encrypting both the note plaintext and the swap plaintext, but with different nonces.